### PR TITLE
fix(52455): destructuring function props with "enum" as property name breaks .d.ts 

### LINKED ISSUE #52455: destructuring function props with "enum" as property name breaks .d.ts
# Bug Report

### 🔎 Search Terms

Enum, declaration, d.ts, renaming of 'enum'

### 🕗 Version & Regression Information

- This is a crash (sort of, .d.ts is "corrupt")
- This changed between versions v4.7.4 and v4.8.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBACgTgezAZygXigbwLACgoFQQB2ArgLYBcUARgggDYQCGxA3HoVAsRNXYxbtOhYAHcE-ek1YdcAXzl4AZqWIBjYAEse3cluABRMuQAUmIierkQxigBooAOhdwIyYFHnV4SZAEosEQI3YFI4Yig3Dzx5IA)

### 💻 Code

```ts
type Props = {
    enum: boolean;
    one: boolean;
};

function omitEnum({ enum: myEnum, ...rest }: Props) {
    return rest
}
```

### 🙁 Actual behavior

`.d.ts` in `v4.8.4` and later

```ts
declare type Props = {
    enum: boolean;
    one: boolean;
    two: boolean;
};
declare function omitEnum({ enum, ...rest }: Props): {
    one: boolean;
    two: boolean;
};
```

Because `enum` isn't re-assigned, TypeScript complains with `':' expected.`

### 🙂 Expected behavior

`.d.ts` in `v4.7.4`

```ts
declare type Props = {
    enum: boolean;
    one: boolean;
    two: boolean;
};
declare function omitEnum({ enum: myEnum, ...rest }: Props): {
    one: boolean;
    two: boolean;
};
```

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #52455

### CODE REVIEW

File: src/compiler/transformers/declarations.ts [hunk 1]
```diff
@@ -695,7 +696,7 @@ export function transformDeclarations(context: TransformationContext) {
             if (elem.kind === SyntaxKind.OmittedExpression) {
                 return elem;
             }
-            if (elem.propertyName && isIdentifier(elem.propertyName) && isIdentifier(elem.name) && !elem.symbol.isReferenced) {
+            if (elem.propertyName && isIdentifier(elem.propertyName) && isIdentifier(elem.name) && !elem.symbol.isReferenced && !identifierToKeywordKind(elem.propertyName)) {
```
Comment (jakebailey) [inline src/compiler/transformers/declarations.ts:699]: I feel like this needs to be `isIdentifierANonContextualKeyword` or something; can you try out a contextual keyword like `abstract` and see what happens?

Comment (weswigham) [reply src/compiler/transformers/declarations.ts:699]: `await` (and maybe `async`) probably also another good one to check.

Comment (weswigham) [reply src/compiler/transformers/declarations.ts:699]: Yeah, none of `async`, `await`, or `abstract` need to be renamed in this context, so definitely need to check if it's a contextual keyword or not.

Comment (a-tarasyuk) [reply src/compiler/transformers/declarations.ts:699]: I think `isIdentifierANonContextualKeyword` doesn't handle `enum`, `function`, etc.

Comment (a-tarasyuk) [reply src/compiler/transformers/declarations.ts]: Ok

**Review by weswigham — APPROVED** [review]

**Pull request merged**